DMCompiler/DM/Expressions/Constant.cs

```diff
@@ -0,0 +1,262 @@
+using OpenDreamShared.Dream;
+using System;
+
+namespace DMCompiler.DM.Expressions {
+    abstract class Constant : DMExpression {
+        public sealed override Constant ToConstant() {
+            return this;
+        }
+
+        public abstract bool IsTruthy();
+
+#region Unary Operations
+        public Constant Not() {
+            return new Number(IsTruthy() ? 0 : 1);
+        }
+
+        public virtual Constant Negate() {
+            throw new Exception($"const operation `-{this}` is invalid");
+        }
+
+        public virtual Constant BinaryNot() {
+            throw new Exception($"const operation `~{this}` is invalid");
+        }
+#endregion
+
+#region Binary Operations
+        public Constant And(Constant rhs) {
+            var truthy = IsTruthy() && rhs.IsTruthy();
+            return new Number(truthy ? 1 : 0);
+        }
+
+        public Constant Or(Constant rhs) {
+            var truthy = IsTruthy() || rhs.IsTruthy();
+            return new Number(truthy ? 1 : 0);
+        }
+
+        public virtual Constant Add(Constant rhs) {
+            throw new Exception($"const operation `{this} + {rhs}` is invalid");
+        }
+
+        public virtual Constant Subtract(Constant rhs) {
+            throw new Exception($"const operation `{this} - {rhs}` is invalid");
+        }
+
+        public virtual Constant Multiply(Constant rhs) {
+            throw new Exception($"const operation `{this} * {rhs}` is invalid");
+        }
+
+        public virtual Constant Divide(Constant rhs) {
+            throw new Exception($"const operation `{this} / {rhs}` is invalid");
+        }
+
+        public virtual Constant Modulo(Constant rhs) {
+            throw new Exception($"const operation `{this} % {rhs}` is invalid");
+        }
+
+        public virtual Constant Power(Constant rhs) {
+            throw new Exception($"const operation `{this} ** {rhs}` is invalid");
+        }
+
+        public virtual Constant LeftShift(Constant rhs) {
+            throw new Exception($"const operation `{this} << {rhs}` is invalid");
+        }
+
+        public virtual Constant RightShift(Constant rhs) {
+            throw new Exception($"const operation `{this} >> {rhs}` is invalid");
+        }
+
+        public virtual Constant BinaryAnd(Constant rhs) {
+            throw new Exception($"const operation `{this} & {rhs}` is invalid");
+        }
+
+        public virtual Constant BinaryXor(Constant rhs) {
+            throw new Exception($"const operation `{this} ^ {rhs}` is invalid");
+        }
+
+        public virtual Constant BinaryOr(Constant rhs) {
+            throw new Exception($"const operation `{this} | {rhs}` is invalid");
+        }
+#endregion
+    }
+
+    // null
+    class Null : Constant {
+        public override void EmitPushValue(DMObject dmObject, DMProc proc) {
+            proc.PushNull();
+        }
+
+        public override bool IsTruthy() => true;
```

Comment:
Should this return false?

Reply:
yeah i think i copied this from a now removed Not() implementation